src/frequenz/channels/util/_periodic_timer.py

```diff
@@ -0,0 +1,201 @@
+        `True`).
+        """
+
+        self._current_drift: timedelta | None = None
+        """The difference between `_next_msg_time` and the actual time when the timer fired.
+
+        This is calculated by `ready()` but is returned by `consume()`. If
+        `None` it means `ready()` wasn't called and `consume()` will assert.
+        `consume()` will set it back to `None` to tell `ready()` that it needs
+        to wait again.
+        """
+
+        if auto_start:
+            self.reset()
+
+    def reset(self) -> None:
+        """Reset the timer to start timing from now.
+
+        If the timer was stopped, or not started yet, it will be started.
+
+        This can only be called with a running loop, see the class
+        documentation for more details.
+        """
+        self._stopped = False
+        self._next_msg_time = asyncio.get_running_loop().time() + self._interval
+
+    def stop(self) -> None:
+        """Stop the timer.
+
+        Once `stop` has been called, all subsequent calls to `ready()` will
+        immediately return False and calls to `consume()` / `receive()` or any
+        use of the async iterator interface will raise
+        a `ReceiverStoppedError`.
+
+        You can restart the timer with `reset()`.
+        """
+        self._stopped = True
+        self._next_msg_time = None
+
+    async def ready(self) -> bool:
+        """Wait until the timer interval passed.
+
+        Once a call to `ready()` has finished, the resulting drift must be read
+        with a call to `consume()` (`receive()` or iterated over) to tell the
+        timer it should wait for the next interval.
+
+        The timer will remain ready (this method will return immediately)
+        until it is consumed.
+
+        Returns:
+            Whether the timer is still running (if it was `stop()`ed, it will
+                return `False`.
+
+        Raises:
+            RuntimeError: if it was called without a running loop.
+        """
+        # if there are messages waiting to be consumed, return immediately.
+        if self._current_drift is not None:
+            return True
+
+        # If a stop was explicitly requested, we bail out.
+        if self._stopped:
+            return False
+
+        # If there is no stop requested but we don't have a time for the next
+        # message, then we reset() (so we start the timer as of now).
+        if self._next_msg_time is None:
+            self.reset()
+        assert self._next_msg_time is not None, "This should be assigned by reset()"
+
+        now = asyncio.get_running_loop().time()
+        diff = self._next_msg_time - now
+        if diff > 0:
+            await asyncio.sleep(diff)
+            now = asyncio.get_running_loop().time()
+        self._current_drift = timedelta(seconds=now - self._next_msg_time)
+        self._next_msg_time = self._next_msg_time + self._interval
+
+        return True
+
+    def consume(self) -> timedelta:
```

Comment:
I didn't get the second part here. About returning `datetime` vs `timedelta` (or both), I know I thought about it and went with `timedelta`, but now I don't remember why.

This is a disadvantage of letting a PR sit there for so long, now I forgot all the context. I will get back to it and probably I will have more comments on your comments :)

Reply:
I was just proposing that we just have a single `utils.Timer` that returns both a periodic time, with uniform increments, and the tick time (i.e) clock time of the tick.

instead of having a separate class for each.  it would be the same performance-wise anyway.

Reply:
I guess I put the delta because for the resampler I didn't need the timestamp, as I when the next resample was supposed to happen, but I guess other users might not be tracking this is it is good information to provide. The drift is used in the resampler to produce warnings if it is too big.

I will change to return `(supposed_fire_time, drift)`, I guess this is the most useful info, when it was supposed to fire (the *logical fire time*) and how it did differ from the actual time it fired. But actually any combination works, as it is possible to derive one from the other. Since our only use case so far has a case for the `(supposed_fire_time, drift)`, I would rather go with that than for what is needed for other hypothetical cases.

I still don't understand what you mean about this:

> Or maybe just two timestamps, and we can just have one timer utils.Timer instead:
>
> ```py
> class TimerTick:
>     timer_period_ts: datetime
>     timer_tick_ts: datetime
> ```

Reply:
Maybe the field names weren't clear.  `utils.Timer` returns `actual_fire_time` only.  `utils.PeriodicTimer` returns `supposed_fire_time/drift`.  We can merge them both,  so there's no `PeriodicTimer`, only `Timer` that provides both types of info.

```python
class TimerTick:
    supposed_fire_time: datetime
    actual_fire_time: datetime
```

Reply:
I see. I don't think it makes sense to merge them though, the way both work is very different, `Timer` will start *ticking* the moment you call `receive()` (or `ready()` technically), while the `PeriodicTimer` needs to be *ticking* all the time once it is started.

For a network timeout, `Timer` makes a lot of sense, so if we make it tick all the time as the periodic timer, we'll have to start/stop it manually to mimic the current behavior, which would make it harder to use.

Reply:
ok, if we don't go the consolidation route, we might want to add a docstring line here explaining the difference from utils.Timer, and a line in utils.Timer saying when people should use a PeriodicTimer instead.

Reply:
Now I remember why I was using only the drift, this is because we are using the monotonic time, so we have the time as a relative `float`, there is no timezone involved here, just an incrementing number. We need to create a datetime if we want to return one. I think we can do it, but since for my use case I didn't need it, I just preferred to go around this, as the user actually has all the information it needs. The user knows at what time the timer started, so it knows when the next tick is supposed to be due.

Also with the leap second it gets tricky, there is the theoretical scenario where as you might have a drift of 0.5, a supposed_fire_time of 12:00:00.00 and an actual_fire_time of 11:59:59.89 (but the drift was positive, it can never fire early). But we said this is rare enough (and it is supposed to go away) so I guess we can ignore it.

I will create the `datetime` anyway, but this adds an extra syscall to get the `datetime.now()` that didn't exit so far.

Reply:
After looking at the code again I'm not that sure it is a good idea to return any `datetime`. The user already have all the info they need to track this themselves **iff** they need it, and I don't think many users of this class will actually need it, if you just want to run something periodically I don't see why we should assume the user cares at all about at what day or time this is triggering. If it were free to return this, I agree we should do it, this class doesn't know anything about `datetime` so far...

Reply:
Do you need to call `datetime.now()` for every tick, if you just want to send the periodic time corresponding to the tick?